sync.go

```diff
@@ -255,36 +317,40 @@ func (lw *LibWallet) RescanBlocks() error {
 		progress := make(chan wallet.RescanProgress, 1)
 		ctx, _ := contextWithShutdownCancel(context.Background())
 
-		var totalHeight int32
+		var totalHeightRescanned int32
 		go lw.wallet.RescanProgressFromHeight(ctx, netBackend, 0, progress)
 
 		for p := range progress {
 			if p.Err != nil {
 				log.Error(p.Err)
-
 				return
 			}
-			totalHeight += p.ScannedThrough
+			totalHeightRescanned += p.ScannedThrough
 			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnRescan(p.ScannedThrough, SyncStateProgress)
+				report := &HeadersRescanProgressReport{
+					CurrentRescanHeight: p.ScannedThrough,
+					TotalHeadersToScan:  lw.GetBestBlock(),
+				}
+				syncProgressListener.OnHeadersRescanProgress(report)
 			}
-		}
 
-		select {
-		case <-ctx.Done():
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnRescan(totalHeight, SyncStateProgress)
-			}
-		default:
-			for _, syncProgressListener := range lw.syncProgressListeners {
-				syncProgressListener.OnRescan(totalHeight, SyncStateFinish)
+			select {
+			case <-ctx.Done():
+				log.Info("Rescan cancelled through context")
+				return
+			default:
+				continue
 			}
 		}
```

Comment:
After the for loop completes, add code here to send final rescan progress notification (to be on the safe side). The current rescan height of this last notification would be equal to the total headers to scan value to signify to the app that rescan has completed.